File: app/src/main/java/com/github/overpass/gather/screen/auth/register/confirm/ConfirmEmailFragment.java
package com.github.overpass.gather.screen.auth.register.confirm;

import androidx.lifecycle.ViewModelProviders;

import com.github.overpass.gather.R;
import com.github.overpass.gather.screen.auth.register.RegistrationFragment;
import com.github.overpass.gather.screen.dialog.ProgressDialogFragment;

import butterknife.OnClick;

import static com.github.overpass.gather.model.commons.UIUtil.toast;

public class ConfirmEmailFragment extends RegistrationFragment<ConfirmEmailViewModel> {

    public static ConfirmEmailFragment newInstance() {
        return new ConfirmEmailFragment();
    }

    @Override
    protected ConfirmEmailViewModel createViewModel() {
        return ViewModelProviders.of(this).get(ConfirmEmailViewModel.class);
    }

    @Override
    protected int getLayoutRes() {
        return R.layout.fragment_confirm_email;
    }

    private void handleConfirmation(ConfirmEmailStatus confirmEmailStatus) {
        switch (confirmEmailStatus.tag()) {
            case ConfirmEmailStatus.SUCCESS:
                handleSuccess(confirmEmailStatus.as(ConfirmEmailStatus.Success.class));
                break;
            case ConfirmEmailStatus.ERROR:
                handleError(confirmEmailStatus.as(ConfirmEmailStatus.Error.class));
                break;
            case ConfirmEmailStatus.PROGRESS:
                handleProgress(confirmEmailStatus.as(ConfirmEmailStatus.Progress.class));
                break;
            default:
                handleFail(confirmEmailStatus.as(ConfirmEmailStatus.Fail.class));
        }
    }

    @OnClick(R.id.tvConfirm)
    public void onConfirmClicked() {
        viewModel.confirm().observe(getViewLifecycleOwner(), this::handleConfirmation);
    }

    private void handleSuccess(ConfirmEmailStatus.Success success) {
        ProgressDialogFragment.hide(getFragmentManager());
        viewModel.setEmailConfirmed();
        viewModel.moveToNextStep(registrationController);
    }

    private void handleError(ConfirmEmailStatus.Error error) {
        ProgressDialogFragment.hide(getFragmentManager());
        String message = error.getThrowable().getLocalizedMessage();
        toast(this, message);
    }

    private void handleProgress(ConfirmEmailStatus.Progress progress) {
        ProgressDialogFragment.show(getFragmentManager());
    }

    private void handleFail(ConfirmEmailStatus.Fail fail) {
        ProgressDialogFragment.hide(getFragmentManager());
        toast(this, fail.getMessage());
    }
}

File: app/src/main/java/com/github/overpass/gather/screen/auth/register/confirm/ConfirmEmailFragment.kt
package com.github.overpass.gather.screen.auth.register.confirm

import android.os.Bundle
import android.view.View
import androidx.lifecycle.Observer
import com.github.overpass.gather.App.Companion.appComponentManager
import com.github.overpass.gather.R
import com.github.overpass.gather.di.register.confirm.ConfirmationComponent
import com.github.overpass.gather.di.register.confirm.ConfirmationComponentManager
import com.github.overpass.gather.model.commons.UIUtil.toast
import com.github.overpass.gather.screen.auth.register.RegistrationFragment
import com.github.overpass.gather.screen.dialog.progress.indeterminate.ProgressDialogFragment
import kotlinx.android.synthetic.main.fragment_confirm_email.*

class ConfirmEmailFragment : RegistrationFragment<ConfirmEmailViewModel, ConfirmationComponent>() {

    override val layoutRes: Int = R.layout.fragment_confirm_email

    override val componentManager: ConfirmationComponentManager =
            appComponentManager.getConfirmationComponentManager()

    override fun createComponent(): ConfirmationComponent = componentManager.getOrCreate(Unit)

    override fun onComponentCreated(component: ConfirmationComponent) {
        component.inject(this)
    }

    override fun createViewModel(): ConfirmEmailViewModel {
        return viewModelProvider.get(ConfirmEmailViewModel::class.java)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        tvConfirm.setOnClickListener {
            viewModel.confirm()
        }
    }

    override fun onBind() {
        super.onBind()
        viewModel.confirmationError().observe(viewLifecycleOwner, Observer { handleError(it) })
        viewModel.confirmationSuccess().observe(viewLifecycleOwner, Observer { handleSuccess() })
        viewModel.confirmationProgress().observe(viewLifecycleOwner, Observer { handleProgress() })
    }

    private fun handleSuccess() {
        ProgressDialogFragment.hide(fragmentManager)
        viewModel.setEmailConfirmed()
        registrationController!!.moveToNextStep()
    }

    private fun handleError(message: String) {
        toast(this, message)
    }

    private fun handleProgress() {
        ProgressDialogFragment.show(fragmentManager)
    }

    companion object {

        @JvmStatic
        fun newInstance(): ConfirmEmailFragment {
            return ConfirmEmailFragment()
        }
    }
}